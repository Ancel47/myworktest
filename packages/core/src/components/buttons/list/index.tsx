import React, { FC } from "react";
import { Button, ButtonProps } from "antd";
import { BarsOutlined } from "@ant-design/icons";
import { useHistory, useParams, useRouteMatch } from "react-router-dom";

import { MatchRoute, ResourceRouterParams } from "../../../interfaces";
import humanizeString from "humanize-string";
import { useTranslate } from "@hooks";

type ListButtonProps = ButtonProps & {
    resourceName?: string;
};

export const ListButton: FC<ListButtonProps> = ({
    resourceName: propResourceName,
    ...rest
}) => {
    const history = useHistory();
    const translate = useTranslate();

    const { resource: routeResourceName } = useParams<ResourceRouterParams>();

    const resourceName = propResourceName ?? routeResourceName;

    return (
        <Button
            onClick={(): void => history.push(`/resources/${resourceName}`)}
            type="default"
            icon={<BarsOutlined />}
            {...rest}
        >
            {translate(
                `common.resources.${resourceName}.title`,
                humanizeString(resourceName),
            )}
        </Button>
    );
};
