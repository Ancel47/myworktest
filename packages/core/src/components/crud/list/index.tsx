import React from "react";
import { Card, Row, Col } from "antd";
import humanizeString from "humanize-string";
import { useParams } from "react-router-dom";

import { useResourceWithRoute, useTranslate } from "@hooks";
import { OptionalComponent } from "@definitions";
import { CreateButton } from "@components";
import { ResourceRouterParams } from "../../../interfaces";

export interface ListProps {
    resourceName: string;
    canCreate?: boolean;
    actionButtons?: React.FC;
    aside?: React.FC;
    title?: string;
    isModalShown?: () => void;
    canExport?: boolean;
}

export const List: React.FC<ListProps> = ({
    canCreate,
    aside,
    title,
    actionButtons,
    children,
    isModalShown,
}) => {
    const { resource: routeResourceName } = useParams<ResourceRouterParams>();

    const resource = useResourceWithRoute(routeResourceName);
    const translate = useTranslate();

    const defaultExtra = canCreate && (
        <CreateButton size="middle" onClick={isModalShown} />
    );

    const renderExtra = () => {
        if (actionButtons) {
            return actionButtons;
        }

        return defaultExtra;
    };

    return (
        <Row gutter={[16, 16]}>
            <Col flex="1 1 200px">
                <Card
                    bodyStyle={{ padding: 0, flex: 1 }}
                    title={
                        title ??
                        translate(
                            `common:resources.${resource.name}.title`,
                            humanizeString(resource.name),
                        )
                    }
                    extra={renderExtra()}
                >
                    {children}
                </Card>
            </Col>

            {aside && (
                <Col flex="0 1 300px">
                    <OptionalComponent optional={aside} />
                </Col>
            )}
        </Row>
    );
};
