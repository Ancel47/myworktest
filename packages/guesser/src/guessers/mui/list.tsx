import * as RefineMui from "@pankod/refine-mui";

import { createGuesser } from "@/create-guesser";
import {
    jsx,
    componentName,
    prettyString,
    accessor,
    printImports,
    toPlural,
    isIDKey,
    dotAccessor,
} from "@/utilities";

import { ErrorComponent } from "./error";
import { LoadingComponent } from "./loading";
import { CodeViewerComponent } from "./code-viewer";

import { GuessField } from "@/types";

/**
 * @experimental This is an experimental component
 */
export const ListGuesser = createGuesser({
    type: "list",
    additionalScope: [["@pankod/refine-mui", "RefineMui", RefineMui]],
    codeViewerComponent: CodeViewerComponent,
    loadingComponent: LoadingComponent,
    errorComponent: ErrorComponent,
    renderer: ({ resource, fields }) => {
        const COMPONENT_NAME = componentName(resource.name, "list");
        const recordName = "dataGridProps?.rows";
        const imports: Array<[element: string, module: string]> = [
            ["useDataGrid", "@pankod/refine-mui"],
            ["DataGrid", "@pankod/refine-mui"],
            ["GridColumns", "@pankod/refine-mui"],
            ["EditButton", "@pankod/refine-mui"],
            ["ShowButton", "@pankod/refine-mui"],
            ["List", "@pankod/refine-mui"],
        ];

        const relationFields: (GuessField | null)[] = fields.filter(
            (field) => field?.relation && !field?.fieldable && field?.resource,
        );

        const relationHooksCode = relationFields
            .filter(Boolean)
            .map((field) => {
                if (field?.relation && !field.fieldable && field.resource) {
                    imports.push(["useMany", "@pankod/refine-core"]);

                    let idsString = "";

                    if (field.multiple) {
                        idsString = `[].concat(...(${recordName}?.map((item) => ${accessor(
                            "item",
                            field.key,
                            field.accessor,
                            false,
                        )}) ?? []))`;
                    } else {
                        idsString = `${recordName}?.map((item) => ${accessor(
                            "item",
                            field.key,
                            field.accessor,
                            false,
                        )}) ?? []`;
                    }

                    return `
                    const { data: ${toPlural(
                        field.resource.name,
                    )}Data, isLoading: ${toPlural(
                        field.resource.name,
                    )}IsLoading } =
                    useMany({
                        resource: "${field.resource.name}",
                        ids: ${idsString},
                        queryOptions: {
                            enabled: !!${recordName},
                        },
                    });
                    `;
                }
                return undefined;
            })
            .filter(Boolean);

        const relationVariableNames = relationFields
            ?.map((field) => {
                if (field && field.resource) {
                    return `${toPlural(field.resource.name)}Data?.data`;
                }
                return undefined;
            })
            .filter(Boolean);

        const renderRelationFields = (field: GuessField) => {
            if (field.relation && field.resource) {
                const variableName = `${toPlural(
                    field.resource.name,
                )}Data?.data`;
                const variableIsLoading = `${toPlural(
                    field.resource.name,
                )}IsLoading`;

                if (Array.isArray(field.accessor)) {
                    // not handled - not possible case
                    return undefined;
                }

                const loadingCondition = `${variableIsLoading} ? <>Loading...</> : `;

                const fieldProperty = `field: "${field.key}"`;

                const valueGetterProperty = field.accessor
                    ? `valueGetter: ({ row }) => {
                    const value = ${accessor(
                        "row",
                        field.key,
                        field.accessor,
                        false,
                    )};

                    return value;
                },`
                    : "";

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = "";

                // if multiple, then map it with tagfield
                // if not, then just show the value

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);
                    let val = "item";

                    // for multiple
                    if (field?.relationGuess) {
                        const valSingle = `${variableName}?.find((resourceItems) => resourceItems.id === item)`;
                        const valViewableSingle = accessor(
                            valSingle,
                            undefined,
                            field?.relationGuess?.accessor,
                        );
                        val = valViewableSingle;
                    }

                    renderCell = `
                    renderCell: function render({ value }) {
                        return ${loadingCondition} (
                            <>
                                {${accessor(
                                    "value",
                                    undefined,
                                    field.accessor,
                                )}?.map((item, index) => (
                                    <TagField key={index} value={${val}} />
                                ))}
                            </>
                        )
                    }
                    `;
                } else {
                    if (field?.relationGuess) {
                        // for single
                        const valSingle = `${variableName}?.find((item) => item.id === value)`;
                        const valViewableSingle = accessor(
                            valSingle,
                            undefined,
                            field?.relationGuess?.accessor,
                        );

                        renderCell = `
                        renderCell: function render({ value }) {
                            return ${loadingCondition} ${valViewableSingle};
                        }
                        `;
                    } else {
                        renderCell = "";
                    }
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        ${valueGetterProperty}
                        minWidth: 300,
                        ${renderCell}
                    }
                `;
            }
            return undefined;
        };

        const imageFields = (field: GuessField) => {
            if (field.type === "image") {
                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                    renderCell: function render({ value }) {
                        return (
                            <img src={${accessor(
                                "value",
                                undefined,
                                Array.isArray(field.accessor)
                                    ? field.accessor
                                    : undefined,
                                " + ",
                            )}} style={{ height: "50px", maxWidth: "100px" }} />
                        )
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    renderCell = `
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item, index) => (
                                    <img src={${val}} key={index} style={{ height: "50px", maxWidth: "100px" }} />
                                ))}
                                </>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        minWidth: 250,
                        ${renderCell}
                    }
                `;
            }
            return undefined;
        };

        const emailFields = (field: GuessField) => {
            if (field.type === "email") {
                imports.push(
                    ["TagField", "@pankod/refine-antd"],
                    ["EmailField", "@pankod/refine-antd"],
                );

                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                renderCell: function render({ value }) {
                    return (
                        <EmailField value={${accessor(
                            "value",
                            undefined,
                            Array.isArray(field.accessor)
                                ? field.accessor
                                : undefined,
                            " + ",
                        )}} />
                    )
                }
            `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    renderCell = `
                    renderCell: function render({ value }) {
                        return (
                            <>
                            {value?.map((item, index) => (
                                <TagField value={${val}} key={index} />
                            ))}
                            </>
                        )
                    }
                `;
                }

                return `
                {
                    ${fieldProperty},
                    ${headerProperty},
                    minWidth: 250,
                    ${renderCell}
                }
            `;
            }
            return undefined;
        };

        const urlFields = (field: GuessField) => {
            if (field.type === "url") {
                imports.push(
                    ["UrlField", "@pankod/refine-mui"],
                    ["TagField", "@pankod/refine-mui"],
                );

                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                    renderCell: function render({ value }) {
                        return (
                            <UrlField value={${accessor(
                                "value",
                                undefined,
                                Array.isArray(field.accessor)
                                    ? field.accessor
                                    : undefined,
                                " + ",
                            )}} />
                        )
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " + ",
                    );

                    renderCell = `
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item, index) => (
                                    <TagField value={${val}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        minWidth: 250,
                        ${renderCell}
                    }
                `;
            }
            return undefined;
        };

        const booleanFields = (field: GuessField) => {
            if (field?.type) {
                imports.push(["Checkbox", "@pankod/refine-mui"]);

                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                    renderCell: function render({ value }) {
                        return (
                            <Checkbox checked={!!${accessor(
                                "value",
                                undefined,
                                Array.isArray(field.accessor)
                                    ? field.accessor
                                    : undefined,
                                " && ",
                            )}} />
                        );
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        " && ",
                    );

                    renderCell = `
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value.map((item, index) => (
                                    <Checkbox checked={!!${val}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        minWidth: 250,
                        ${renderCell}
                    }
                `;
            }

            return undefined;
        };

        const dateFields = (field: GuessField) => {
            if (field.type === "date") {
                imports.push(["DateField", "@pankod/refine-mui"]);

                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                    renderCell: function render({ value }) {
                        return <DateField value={value} />;
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        ' + " " + ',
                    );
                    renderCell = `
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item, index) => (
                                    <DateField value={${val}} key={index} />
                                ))}
                                </>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        minWidth: 250,
                        ${renderCell}
                    }
                `;
            }
            return undefined;
        };

        const richtextFields = (field: GuessField) => {
            if (field?.type === "richtext") {
                imports.push(["MarkdownField", "@pankod/refine-mui"]);

                const fieldProperty =
                    Array.isArray(field.accessor) || field.multiple
                        ? `field: "${field.key}"`
                        : field.accessor
                        ? `field: "${dotAccessor(
                              field.key,
                              undefined,
                              field.accessor,
                          )}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = `
                    renderCell: function render({ value }) {
                        return <MarkdownField value={(value ?? "").slice(0, 80) + "..."} />;
                    }
                `;

                if (field.multiple) {
                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        ' + " " + ',
                    );
                    renderCell = `
                        renderCell: function render({ value }) {
                            return (
                                <>
                                {value?.map((item, index) => (
                                    <MarkdownField value={(${val}).slice(0, 80) + "..."} key={index} />
                                ))}
                                </>
                            )
                        }
                    `;
                }

                return `
                    {
                        ${fieldProperty},
                        ${headerProperty},
                        minWidth: 250,
                        ${renderCell}
                    }
                `;
            }

            return undefined;
        };

        const basicFields = (field: GuessField) => {
            if (field && (field.type === "text" || field.type === "number")) {
                const fieldProperty =
                    field.accessor &&
                    !Array.isArray(field.accessor) &&
                    !field.multiple
                        ? `field: "${field.key}.${field.accessor}"`
                        : `field: "${field.key}"`;

                const headerProperty = `headerName: "${prettyString(
                    field.key,
                )}"`;

                let renderCell = "";

                if (field.multiple) {
                    imports.push(["TagField", "@pankod/refine-mui"]);

                    const val = accessor(
                        "item",
                        undefined,
                        field.accessor,
                        ' + " " + ',
                    );
                    renderCell = `
                    renderCell: function render({ row }) {
                        return (
                            <>
                                {${accessor("row", field.key)}.map((item) => (
                                    <TagField value={${val}} key={${val}} />
                                ))}
                            </>
                        )
                    }
                    `;
                }
                if (!field.multiple && Array.isArray(field.accessor)) {
                    renderCell = `
                    renderCell: function render({ row }) {
                        return 
                            <>{${accessor("row", field.key, field.accessor)}}</>
                        );
                    }
                    `;
                }

                return `
                {
                    ${fieldProperty},
                    ${headerProperty},
                    ${field.type === "number" ? "type: 'number'," : ""}
                    minWidth: ${isIDKey(field.key) ? 50 : 200},
                    ${renderCell}
                }
                `;
            }
            return undefined;
        };

        const actionButtons = `
                {
                    field: "actions",
                    headerName: "Actions",
                    renderCell: function render({ row }) {
                        return (
                            <>
                                <EditButton hideText recordItemId={row.id} />
                                <ShowButton hideText recordItemId={row.id} />
                            </>
                        );
                    },
                    align: "center",
                    headerAlign: "center",
                    minWidth: 80,
                },
        `;

        const renderedFields: Array<string | undefined> = fields.map(
            (field) => {
                switch (field?.type) {
                    case "text":
                    case "number":
                        return basicFields(field);
                    case "richtext":
                        return richtextFields(field);
                    case "email":
                        return emailFields(field);
                    case "image":
                        return imageFields(field);
                    case "date":
                        return dateFields(field);
                    case "boolean":
                        return booleanFields(field);
                    case "url":
                        return urlFields(field);
                    case "relation":
                        return renderRelationFields(field);
                    default:
                        return undefined;
                }
            },
        );

        return jsx`
        ${printImports(imports)}
        
        export const ${COMPONENT_NAME} = () => {
            const { dataGridProps } = useDataGrid();
        
            ${relationHooksCode}

            const columns = React.useMemo<GridColumns<any>>(() => [
                ${[...renderedFields, actionButtons].join(",\r\n")}
            ], [${relationVariableNames.join(",")}]);

            return (
                <List>
                    <DataGrid {...dataGridProps} columns={columns} autoHeight />
                </List>
            );
        };
        `;
    },
});
