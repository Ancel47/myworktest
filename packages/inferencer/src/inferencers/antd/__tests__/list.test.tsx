import React from "react";
import renderer, { act } from "react-test-renderer";
import { Route, Routes } from "react-router-dom";

import { TestWrapper } from "@test";
import { ListInferencer } from "../list";

describe("AntdListInferencer", () => {
    it("should match the snapshot", async () => {
        let node: any;

        await act(async () => {
            const Wrapper = TestWrapper({
                routerInitialEntries: ["/posts"],
                resources: [
                    {
                        name: "posts",
                        list: ListInferencer,
                    },
                    {
                        name: "categories",
                    },
                    {
                        name: "tags",
                    },
                    { name: "users" },
                ],
            });

            node = renderer.create(
                <Wrapper>
                    <Routes>
                        <Route
                            path="/:resource"
                            element={<ListInferencer resource="posts" />}
                        />
                    </Routes>
                </Wrapper>,
            );
        });

        // wait for 5 seconds
        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 5000));
        });

        const tree = node?.toJSON();

        expect(tree).toMatchSnapshot();
    });
});
