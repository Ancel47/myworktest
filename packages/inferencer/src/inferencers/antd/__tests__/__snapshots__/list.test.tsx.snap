// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AntdListInferencer should match the snapshot 1`] = `
Array [
  <div>
    <div>
      <div
        className="ant-page-header has-breadcrumb"
      >
        <div
          className="ant-page-header-heading"
        >
          <div
            className="ant-page-header-heading-left"
          >
            <span
              className="ant-page-header-heading-title"
              title="Posts"
            >
              Posts
            </span>
          </div>
        </div>
        <div
          className="ant-page-header-content"
        >
          <div>
            <div
              className="ant-table-wrapper"
            >
              <div
                className="ant-spin-nested-loading"
              >
                <div>
                  <div
                    aria-busy={true}
                    aria-live="polite"
                    className="ant-spin ant-spin-spinning"
                  >
                    <span
                      className="ant-spin-dot ant-spin-dot-spin"
                    >
                      <i
                        className="ant-spin-dot-item"
                      />
                      <i
                        className="ant-spin-dot-item"
                      />
                      <i
                        className="ant-spin-dot-item"
                      />
                      <i
                        className="ant-spin-dot-item"
                      />
                    </span>
                  </div>
                </div>
                <div
                  className="ant-spin-container ant-spin-blur"
                >
                  <div
                    className="ant-table ant-table-empty ant-table-scroll-horizontal"
                  >
                    <div
                      className="ant-table-container"
                    >
                      <div
                        className="ant-table-content"
                        onScroll={[Function]}
                        style={
                          Object {
                            "overflowX": "auto",
                            "overflowY": "hidden",
                          }
                        }
                      >
                        <table
                          style={
                            Object {
                              "minWidth": "100%",
                              "tableLayout": "auto",
                              "width": "auto",
                            }
                          }
                        >
                          <colgroup />
                          <thead
                            className="ant-table-thead"
                          >
                            <tr>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Id
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Title
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Slug
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Content
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Hit
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Category
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                User
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Status
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Status Color
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Created At
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Published At
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Image
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Tags
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Language
                              </th>
                              <th
                                className="ant-table-cell"
                                colSpan={null}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                Actions
                              </th>
                            </tr>
                          </thead>
                          <tbody
                            className="ant-table-tbody"
                          >
                            <tr
                              aria-hidden="true"
                              className="ant-table-measure-row"
                              style={
                                Object {
                                  "fontSize": 0,
                                  "height": 0,
                                }
                              }
                            >
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                              <td
                                style={
                                  Object {
                                    "border": 0,
                                    "height": 0,
                                    "padding": 0,
                                  }
                                }
                              >
                                <div
                                  style={
                                    Object {
                                      "height": 0,
                                      "overflow": "hidden",
                                    }
                                  }
                                >
                                   
                                </div>
                              </td>
                            </tr>
                            <tr
                              className="ant-table-placeholder"
                              style={
                                Object {
                                  "display": null,
                                }
                              }
                            >
                              <td
                                className="ant-table-cell"
                                colSpan={15}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                rowSpan={null}
                                style={Object {}}
                              >
                                <div
                                  className="ant-table-expanded-row-fixed"
                                  style={
                                    Object {
                                      "left": 0,
                                      "overflow": "hidden",
                                      "position": "sticky",
                                      "width": 0,
                                    }
                                  }
                                />
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    style={
      Object {
        "bottom": 12,
        "position": "fixed",
        "right": 60,
      }
    }
  >
    <div
      className="ant-space ant-space-horizontal ant-space-align-center"
      style={Object {}}
    >
      <div
        className="ant-space-item"
        style={
          Object {
            "marginRight": 8,
          }
        }
      >
        <button
          className="ant-btn ant-btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span
            aria-label="code"
            className="anticon anticon-code"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="code"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M516 673c0 4.4 3.4 8 7.5 8h185c4.1 0 7.5-3.6 7.5-8v-48c0-4.4-3.4-8-7.5-8h-185c-4.1 0-7.5 3.6-7.5 8v48zm-194.9 6.1l192-161c3.8-3.2 3.8-9.1 0-12.3l-192-160.9A7.95 7.95 0 00308 351v62.7c0 2.4 1 4.6 2.9 6.1L420.7 512l-109.8 92.2a8.1 8.1 0 00-2.9 6.1V673c0 6.8 7.9 10.5 13.1 6.1zM880 112H144c-17.7 0-32 14.3-32 32v736c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V144c0-17.7-14.3-32-32-32zm-40 728H184V184h656v656z"
              />
            </svg>
          </span>
          <span>
            Show Code
          </span>
        </button>
      </div>
      <div
        className="ant-space-item"
        style={Object {}}
      >
        <a
          className="ant-btn ant-btn-default"
          href="https://github.com/refinedev/refine/issues"
          onClick={[Function]}
          target="_blank"
        >
          <span
            aria-label="message"
            className="anticon anticon-message"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="message"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M464 512a48 48 0 1096 0 48 48 0 10-96 0zm200 0a48 48 0 1096 0 48 48 0 10-96 0zm-400 0a48 48 0 1096 0 48 48 0 10-96 0zm661.2-173.6c-22.6-53.7-55-101.9-96.3-143.3a444.35 444.35 0 00-143.3-96.3C630.6 75.7 572.2 64 512 64h-2c-60.6.3-119.3 12.3-174.5 35.9a445.35 445.35 0 00-142 96.5c-40.9 41.3-73 89.3-95.2 142.8-23 55.4-34.6 114.3-34.3 174.9A449.4 449.4 0 00112 714v152a46 46 0 0046 46h152.1A449.4 449.4 0 00510 960h2.1c59.9 0 118-11.6 172.7-34.3a444.48 444.48 0 00142.8-95.2c41.3-40.9 73.8-88.7 96.5-142 23.6-55.2 35.6-113.9 35.9-174.5.3-60.9-11.5-120-34.8-175.6zm-151.1 438C704 845.8 611 884 512 884h-1.7c-60.3-.3-120.2-15.3-173.1-43.5l-8.4-4.5H188V695.2l-4.5-8.4C155.3 633.9 140.3 574 140 513.7c-.4-99.7 37.7-193.3 107.6-263.8 69.8-70.5 163.1-109.5 262.8-109.9h1.7c50 0 98.5 9.7 144.2 28.9 44.6 18.7 84.6 45.6 119 80 34.3 34.3 61.3 74.4 80 119 19.4 46.2 29.1 95.2 28.9 145.8-.6 99.6-39.7 192.9-110.1 262.7z"
              />
            </svg>
          </span>
          <span>
            Give Feedback
          </span>
        </a>
      </div>
    </div>
  </div>,
]
`;
